machine_learning/linear_regression.py

```diff
@@ -70,8 +69,8 @@ def run_linear_regression(data_x, data_y):
     :param data_y  : contains the output (result vector)
     :return        : feature for line of best fit (Feature vector)
     """
-    iterations = 100000
-    alpha = 0.0001550
+    iterations = 250000
+    alpha = 0.000326
```

Comment:
This won't give the optimal coefficients. Like I said in my original issue, gradient descent is an approximation algorithm _by definition,_ so tweaking the parameters will only give you a better approximation. My point with the original issue is that we should add a _different_ algorithm, one that is a _direct method_ rather than an _iterative method._

Reply:
This also doesn't fix the fact that the code is still calculating SSE incorrectly.

Reply:
can you elaborate on what do you mean by direct method?

Reply:
A [direct method](https://en.wikipedia.org/wiki/Iterative_method) gives an exact solution. By contrast, an iterative method (such as gradient descent) gives increasingly accurate approximations. With an iterative method, you get _closer and closer_ to the exact solution but you may never actually reach that exact solution.

For example, let's say you want to find the roots of a quadratic polynomial $ax^2 + bx + c$. You could use Newton's method to approximate the roots, refining your solutions to be more and more accurate—this is an iterative method. Alternatively, you could just use the quadratic formula to just get the exact solutions—this is a direct method.

Reply:
also, explain with facts why the SSE is calculated incorrectly because the code does the predictions finds the deviation and takes the squares of the deviations, and sums the errors, which is the definition of SSE.

Reply:
I already explained this with my code sample in the original issue, but sure, let's go over the math for the SSE function:
https://github.com/TheAlgorithms/Python/blob/9e08c7726dee5b18585a76e54c71922ca96c0b3a/machine_learning/linear_regression.py#L52-L64

Likes 60–62 are fine because they calculate the SSE exactly as we expect:
$$\mathrm{SSE} = \sum_{i = 0}^{N} (\mathbf{x}_i^{\top} \boldsymbol{\theta} - \mathbf{y}_i)^2$$

However, for some reason, the function then divides the SSE by `2 * len_data`. Dividing the SSE (_sum_ of squared errors) by the number of datapoints gives you the MSE (_mean_ squared error):
$$\mathrm{MSE} = \frac{1}{N} \sum_{i = 0}^{N} (\mathbf{x}_i^{\top} \boldsymbol{\theta} - \mathbf{y}_i)^2$$

Naturally, it follows that the "SSE" function is actually calculating half of the MSE because it's dividing the SSE by $2N$.

Reply:
i appreciate your response